client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,201 @@
+##############################################################################
+# This file contains utility functions for use by the interface:
+# - Creating the mock file system for the user
+# - Timing each task
+# - Determining which task and treatment ordering for the current experiment
+# - Determine the next task to move onto and whether the experiment is over
+# - Verify the output of a task
+##############################################################################
+
+# Converts a numeric ASCII value to a character.
+chr() {
+  [ "$1" -lt 256 ] || return 1
+  printf "\\$(printf '%03o' "$1")"
+}
+
+# Converts a character to its ASCII value.
+ord() {
+  LC_CTYPE=C printf '%d' "'$1"
+}
+
+# Outputs the nth alphabetic character. n is 1-based; that is, char_from(1) is "a".
+# $1: the number n specifying which characters
+char_from() {
+  local num_a=$(ord "a")
+  local num_fr=$((num_a + $1 - 1))
+
+  echo $(chr ${num_fr})
+}
+
+# Gets the true task code from the user task number and the task set.
+# The user task number is always sequential.
+# The task can be in either task set 1 or 2, with the dividing task being
+# TASKS_SIZE / 2.
+# That is, if the current task set is 2 and the user current user task number
+# is 12, then its true task code is "a", because it is in task set 1.
+get_task_code() {
+  if ((task_set == 1)); then
```

Comment:
I found this logic hard to follow.

One possibly-simpler way to think about this is that there are these cases:
 * Second character of treatment = "1":  nothing to do
 * Second character of treatment = "2":  need to add or remove TASKS_SIZE/2.

Or, maybe better documentation of the current logic.

Also, please write test cases for the functions in this file, such as chr, ord, char_from, get_task_code.  Put those test cases in a separate file.

Reply:
Commit 2d725d3d has added a `test` directory under `.infrastructure` which uses Bats to perform the tests. I would really like to hear your thoughts on this.

Over the weekend I will add several spec tests to ensure that all infrastructure code works as well (`next_task`, `precmd`, `preexec`, etc.). I believe the spec tests will help us in the long run as well.

With the addition of the tests, I found out that `get_task_code` does not handle an odd number of tasks (i.e. `$TASKS_SIZE` is odd) very well. The two tests `get_task_code correct - ... odd task size` are skipped because of this reason.

Reply:
That seems fine.

One concern is that this requires another package, so anyone developing on it needs to install bats.  The dependency must not leak out to experimental subjects.  It probably won't, but it's just one more concern.

Another concern with f5845dd is that it's hard to figure out where the tests are and what they do, and they depend on new code written in the tests.  I had expected about  two lines of test for each of the bash functions.  If the function works for 'a' and 'q', then it's hard to imagine that it would not work for all inputs, and it's easy to manually validate the tests and better understand the function's operation if it was not clear from the documentation.  These tests don't do either of those things.  So, I feel that they are overkill.

That said, it's great that adding the tests revealed a limitation.  Maybe the Makefile task that creates the zip file should fail if there are an odd number of tasks.